calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
+
+	iface, addr, err := autodetection.FilteredEnumeration(incl, excl, version)
+	if err != nil {
+		message("Unable to auto-detect any valid interfaces: %s", err)
+		return nil, nil
+	}
+
+	if addr == nil {
+		message("Unable to auto-detect an IPv%d address", version)
+		return nil, nil
+	}
+
+	message("Using autodetected IPv%d address on interface %s: %s", version, iface.Name, addr.IPAddress)
+
+	if addr.IPNetwork != nil {
+		message("Unable to auto-detect an IPv%d network", version)
+	} else {
+		message("Using autodetected IPv%d network on interface %s: %s", version, iface.Name, addr.IPNetwork)
+	}
+
+	return addr.IPAddress, addr.IPNetwork
+}
+
+// configureASNumber configures the Node resource with the AS numnber specified
+// in the environment.
+func configureASNumber(node *api.Node) {
+	// Extract the AS number from the environment
+	asStr := os.Getenv("AS")
+	if asStr != "" {
+		if asNum, err := numorstring.ASNumberFromString(asStr); err != nil {
+			fatal("The AS number specified in the environment is not valid: %s", asStr)
+			terminate()
+		} else {
+			message("Using AS number specified in environment: %s", asNum)
+			node.Spec.BGP.ASNumber = &asNum
+		}
+	} else {
+		message("AS number not specified in environment, using current value")
+	}
+}
+
+// configureIPPools ensures that default IP pools are created (unless explicitly
+// requested otherwise).
+func configureIPPools(client *client.Client) {
+	if strings.ToLower(os.Getenv("NO_DEFAULT_POOLS")) == "true" {
+		log.Info("Not required to configured IP pools")
 		return
 	}
-	log.Infof("Assigned ClusterGUID %s", guid)
+
+	// Get a list of all IP Pools
+	poolList, err := client.IPPools().List(api.IPPoolMetadata{})
+	if err != nil {
+		fatal("Unable to fetch IP pool list: %s", err)
+		terminate()
+	}
+
+	// Check for IPv4 and IPv6 pools and filter for IPIP pools.
+	ipv4Present := false
+	ipv6Present := false
+	for _, p := range poolList.Items {
+		version := p.Metadata.CIDR.Version()
+		ipv4Present = ipv4Present || (version == 4)
+		ipv6Present = ipv6Present || (version == 6)
+		if ipv4Present && ipv6Present {
+			break
+		}
+	}
+
+	// Ensure there are pools created for each IP version.
+	if !ipv4Present {
+		createIPPool(client, DEFAULT_IPV4_POOL_CIDR)
+	}
+	if !ipv6Present && ipv6Supported() {
+		createIPPool(client, DEFAULT_IPV6_POOL_CIDR)
+	}
+}
+
+// ipv6Supported returns true if IPv6 is supported on this platform.  This performs
+// a simplistic check of /proc/sys/net/ipv6 since platforms that do not have IPv6
+// compiled in will not have this entry.
+func ipv6Supported() bool {
+	_, err := os.Stat("/proc/sys/net/ipv6")
+	supported := (err == nil)
+	log.Infof("IPv6 supported on this platform: %v", supported)
+	return supported
+}
+
+// createIPPool creates an IP pool using the specified CIDR string.
+func createIPPool(client *client.Client, cs string) {
+	_, cidr, _ := net.ParseCIDR(cs)
+	version := cidr.Version()
+
+	log.Info("Creating default IPv%d pool", version)
```

Comment:
Not sure I'm too worried here.  We only call this function if we think we need to create a pool.  The only time we won't is when two nodes are concurrently trying to create the pool and the other node gets there first.   So in general we won't log this every time a node starts.